src/signTxPoolRegistration.c

```diff
@@ -0,0 +1,836 @@
+				VALIDATE(view_remainingSize(&view) >= 1, ERR_INVALID_DATA);
+				uint8_t includeIpv4 = parse_u1be(&view);
+				if (includeIpv4 == RELAY_YES) {
+					ipv4Size = 4;
+					VALIDATE(view_remainingSize(&view) >= ipv4Size, ERR_INVALID_DATA);
+					os_memmove(ipv4, view.ptr, ipv4Size);
+					view_skipBytes(&view, ipv4Size);
+					TRACE("ip v4");
+					TRACE_BUFFER(ipv4, ipv4Size);
+					ipv4Ptr = ipv4;
+				} else {
+					VALIDATE(includeIpv4 == RELAY_NO, ERR_INVALID_DATA);
+				}
+			}
+
+			uint8_t ipv6[16];
+			uint8_t* ipv6Ptr = NULL;
+			size_t ipv6Size = 0;
+			{
+				VALIDATE(view_remainingSize(&view) >= 1, ERR_INVALID_DATA);
+				uint8_t includeIpv6 = parse_u1be(&view);
+				if (includeIpv6 == RELAY_YES) {
+					ipv6Size = 16;
+					VALIDATE(view_remainingSize(&view) >= ipv6Size, ERR_INVALID_DATA);
+					os_memmove(ipv6, view.ptr, ipv6Size);
+					view_skipBytes(&view, ipv6Size);
+					TRACE("ip v6");
+					TRACE_BUFFER(ipv6, ipv6Size);
+					ipv6Ptr = ipv6;
+				} else {
+					VALIDATE(includeIpv6 == RELAY_NO, ERR_INVALID_DATA);
+				}
+			}
+
+			VALIDATE(view_remainingSize(&view) == 0, ERR_INVALID_DATA);
+
+			TRACE("Adding relay format 0 to tx hash");
+			txHashBuilder_addPoolRegistrationCertificate_addRelay0(
+			        txHashBuilder, portPtr,
+			        ipv4Ptr, ipv4Size,
+			        ipv6Ptr, ipv6Size
+			);
+			break;
+		}
+
+		case RELAY_SINGLE_HOST_NAME: {
+			uint16_t port;
+			uint16_t *portPtr = NULL;
+			{
+				VALIDATE(view_remainingSize(&view) >= 1, ERR_INVALID_DATA);
+				uint8_t includePort = parse_u1be(&view);
+				if (includePort == RELAY_YES) {
+					port = parse_u2be(&view);
+					TRACE("Port: %u", port);
+					portPtr = &port;
+				} else {
+					VALIDATE(includePort == RELAY_NO, ERR_INVALID_DATA);
+				}
+			}
+
+			VALIDATE(view_remainingSize(&view) <= 64, ERR_INVALID_DATA);
+			str_validateText(VIEW_REMAINING_TO_TUPLE_BUF_SIZE(&view));
+
+			TRACE("Adding relay format 1 to tx hash");
+			txHashBuilder_addPoolRegistrationCertificate_addRelay1(
+			        txHashBuilder,
+			        portPtr,
+			        VIEW_REMAINING_TO_TUPLE_BUF_SIZE(&view)
+			);
+			break;
+		}
+
+		case RELAY_MULTIPLE_HOST_NAME: {
+			VALIDATE(view_remainingSize(&view) <= 64, ERR_INVALID_DATA);
+			str_validateText(VIEW_REMAINING_TO_TUPLE_BUF_SIZE(&view));
+
+			TRACE("Adding relay format 2 to tx hash");
+			txHashBuilder_addPoolRegistrationCertificate_addRelay2(
+			        txHashBuilder,
+			        VIEW_REMAINING_TO_TUPLE_BUF_SIZE(&view)
+			);
+			break;
+		}
+		}
+	}
+
+	respondSuccessEmptyMsg();
+
+	{
+		subctx->currentRelay++;
+		TRACE("current relay %d", subctx->currentRelay);
+
+		if (subctx->currentRelay == subctx->numRelays) {
+			advanceState();
+		}
+	}
+}
+
+
+// ============================== METADATA ==============================
+
+enum {
+	HANDLE_METADATA_STEP_DISPLAY_URL = 6340,
+	HANDLE_METADATA_STEP_DISPLAY_HASH,
+	HANDLE_METADATA_STEP_RESPOND,
+	HANDLE_METADATA_STEP_INVALID,
+};
+
+static void handleMetadata_ui_runStep()
```

Comment:
It probably makes sense to let the user know if they are signing a private pool (i.e. with null metadata), at least that's what we are doing on Trezor